onnxruntime/core/optimizer/common_subexpression_elimination.cc

```diff
@@ -0,0 +1,408 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "common_subexpression_elimination.h"
+#include "core/graph/graph_utils.h"
+
+#include <type_traits>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace onnxruntime {
+
+namespace {
+  struct DeepPointerEquality {
+    template<typename Ptr>
+    bool operator()(const Ptr& lhs, const Ptr& rhs) const {
+      if (lhs == nullptr || rhs == nullptr) {
+        return lhs == rhs;
+      }
+      return *lhs == *rhs;
+    }
+  };
+
+  struct DeepPointerHash {
+    template<typename Ptr>
+    std::size_t operator()(const Ptr& value) const {
+      if (value == nullptr) {
+        return 0;
+      }
+      return std::hash<std::decay_t<decltype(*value)>>{}(*value);
+    }
+  };
+
+  template<typename T, typename Hasher>
+  void UpdateHash(const T& x, Hasher hasher, std::size_t& hash) {
+    constexpr std::size_t kPrime = 31013;
+    hash = hash * kPrime + hasher(x);
+  }
+
+  template<typename T>
+  void UpdateHash(const T& x, std::size_t& hash) {
+    UpdateHash(x, std::hash<T>{}, hash);
+  }
+
+  template<typename Container>
+  void UpdateHashWithContainer(const Container& c, std::size_t& hash) {
+    for (const auto& element : c) {
+      UpdateHash(element, hash);
+    }
+  }
+
+  using OutputIndex = int;
+
+  constexpr OutputIndex kInvalidOutputIndex = -1;
+
+  const NodeArg* Normalize(const NodeArg* node_arg) {
+    return node_arg == nullptr || !node_arg->Exists() ? nullptr : node_arg;
+  }
+
+  // Represents an equivalence class of expressions (inputs, constant initializers and node outputs)
+  // that will always evaluate to the same runtime value.
+  class EquivalenceClass {
+  public:
+    bool operator==(const EquivalenceClass& other) const;
+    bool operator!=(const EquivalenceClass& other) const;
+
+    friend struct ::std::hash<EquivalenceClass>;
+    friend std::vector<std::vector<const EquivalenceClass*>> Normalize(const Node* node, const std::vector<const EquivalenceClass*>& inputs);
+
+    explicit EquivalenceClass(const NodeArg* non_op_value)
+      : attributes_(nullptr)
+      , output_index_(kInvalidOutputIndex)
+      , non_op_value_(Normalize(non_op_value))
+      , discriminator_(0)
+      , hash_(CalculateHash())
+    { }
+
+    EquivalenceClass(const Node* node, const std::vector<const EquivalenceClass*>& explicit_inputs,
+                OutputIndex output_index, int discriminator)
+      : op_type_(node->OpType())
+      , domain_(node->Domain())
+      , inputs_(Normalize(node, explicit_inputs))
+      , attributes_(&node->GetAttributes())
+      , output_index_(output_index)
+      , non_op_value_(nullptr)
+      , discriminator_(discriminator)
+      , hash_(CalculateHash())
+    { }
+
+  private:
+    std::size_t CalculateHash() const;
+
+    // Operation and domain of the node that produces this value.
+    const std::string op_type_;
+    const std::string domain_;
+
```

Comment:
could be references

Reply:
These values are set to empty strings in one of the constructors (when the equivalence class represents an input or constant initializer). So they can't be references.